filter.c

```diff
@@ -1168,6 +1176,135 @@ static int func_avg(filter_t *flt, bcf1_t *line, token_t *rtok, token_t **stack,
     }
     return 1;
 }
+static int compare_doubles(const void *lhs, const void *rhs)
+{
+    double arg1 = *(const double*) lhs;
+    double arg2 = *(const double*) rhs;
+    if (arg1 < arg2) return -1;
+    if (arg1 > arg2) return 1;
+    return 0;
+}
+static double median(double *vals, const int n)
+{
+    switch (n) {
+        case 0: return 0;
+        case 1: return vals[0];
+        case 2: return (vals[0] + vals[1]) / 2;
+        default:
+        {
+            qsort(vals, n, sizeof(double), compare_doubles);
+            const int middle = n / 2;
+            if (n % 2 == 1) {
+                return vals[middle];
+            } else {
+                return (vals[middle - 1] + vals[middle]) / 2;
+            }
+        }
+    }
+}
+static int func_median(filter_t *flt, bcf1_t *line, token_t *rtok, token_t **stack, int nstack)
+{
+    token_t *tok = stack[nstack - 1];
+    rtok->nvalues = 0;
+    if ( !tok->nvalues ) return 1;
+    double *vals = (double*) malloc(tok->nvalues * sizeof(double));
+    int i, n = 0;
+    for (i=0; i<tok->nvalues; i++)
+        if ( !bcf_double_is_missing(tok->values[i]) ) { vals[n] = tok->values[i]; n++; }
+    if ( n )
+    {
+        rtok->values[0] = median(vals, n);
+        rtok->nvalues   = 1;
+    }
+    free(vals);
+    return 1;
+}
+static double mode(double *vals, const int n)
+{
+    switch (n) {
+        case 0: return 0;
+        case 1: return vals[0];
+        case 2: return vals[0] < vals[1] ? vals[0] : vals[1];
+        default:
+        {
+            qsort(vals, n, sizeof(double), compare_doubles);
+            // discretize real values
+            int *int_vals = (int*) malloc(n * sizeof(int));
+            int i;
+            for (i=0; i<n; i++) int_vals[i] = vals[i];
+            // find mode of discrete samples
+            int mode_count = 0, mode_idx = 0;
+            for (i=0; i<n; i++) {
+                int j;
+                for (j = i + 1; j < n && vals[i] == vals[j]; ++j);
+                int count = j - i;
```

Comment:
You could speed up the algorithm by incrementing `i` with `count` at the end of the outer loop, so that you skip the values from the current group.

Reply:
Addressed in d985011bb63955b7e1d6a8e8bda6427d41d30e91.

Reply:
If `mode` is supposed to return the most frequent value, I suggest using `khash` instead, sorting values is not necessary and would not scale well when used on FORMAT fields in VCFs with many samples.

Reply:
I don't have any familiarity with the `k` library so I'm not able to make this change. Another option - if you don't want to use `qsort` - may be to use a radix sort or bucket sort. If the values are actually integers, which would probably be the most common case as the mode of real values is not very useful, then bucket sort would likely be the most efficient option (if the maximum absolute value is not too big).

Reply:
For this problem most efficient is hashing, any type of sort will be slower. We could add it after merging the pull request.

However, I am not completely convinced about adding the `mode` function. It leads to complications for floats, the outcome depends on discretization details. Do you have a practical use case in mind?

Reply:
Hashing is unlikely to be faster than bucket sort for integers values since bucket sort is essentially a hashing where the value is used as the hash. If the domain is large then the cost of allocating the array might outweigh the cost of computing the hash and managing allocations for the hash table. In any case, probably the difference will be small in practice and hashing is more generally applicable so I agree that it is probably the way to go, although I'd be curious to see some benchmarks against `qsort` to see the difference in practice.

As for why the mode is useful. For categorical values it is the only average the makes much sense (e.g. Boolean flag).

Reply:
Sort algorithms have time complexity O(n log(n)) on average and some of them O(n^2) in the worst case whereas hash in O(n). Although sort can be faster for small in some situations, it does not scale up. 

More importantly though: Do you have a real usage scenario where this feature would be useful? Given that the range of a float tag can be about anything (e.g. between [0,1]) I am not convinced `mode` is that useful. It would have to be parametrized by the user, but is that worth supporting.

Reply:
[Counting sort](https://en.wikipedia.org/wiki/Counting_sort) - the variant of bucket sort that I am referring to - is worst-case O(n) time and space complexity.

I'm assuming that all numeric tag types are converted to `double` for the purpose of filter expressions? I agree that mode is not useful for real type input, but if the actual tag type was integral then the mode is useful, and as I mentioned is the only average that makes much sense for categorical types. For example, suppose I define a categorical tag where `0 = red`, `1 = blue`, `2 = green`, then the mean or median of a range of such values is meaningless, but the mode makes a lot of sense. I'm not sure what you mean by "It would have to be parametrized by the user".

Reply:
I understand the concept. The question was aimed at did you add this to your pull request for the sake of completeness or do you have a real use case?